CACI-MIlMOVE/mymove
# B 16771 okta login

# There are four tickets in this pull request
## [B-16771](https://www13.v1host.com/USTRANSCOM38/story.mvc/Summary?oidToken=Story%3a808920) & [B-17148](https://www13.v1host.com/USTRANSCOM38/story.mvc/Summary?oidToken=Story%3a824711) & [B-17153](https://www13.v1host.com/USTRANSCOM38/story.mvc/Summary?oidToken=Story%3a824716) & [B-17159](https://www13.v1host.com/USTRANSCOM38/story.mvc/Summary?oidToken=Story%3a825453)

## Summary

Major changes

- Login gov handlers switched to Okta.
- Login gov tests replaced with Okta tests.
- Login gov providers replaced with Okta goth providers.
- Custom Okta provider wrapping for added functionality.
- API URLs changes from login-gov to Okta.
- Login gov auth has been replaced with Okta auth.
- Introduced http mock tests that can be utilized for RESTful backend testing.
- New libraries provided. This will be detailed in the minor mod.
- Added additional API call inside of callback handling for added security and code exchanging between Okta.
  - It is incredibly unlikely this would impact any rate limiting due to our current high amount allotted.

## Be aware:
Full login is not yet supported as of yet until Okta sessions have been implemented. 

## Verification Steps for Reviewers

- Successful Okta login
- Only error on login is unable to find identity
- Internal server error after redirect (Remember, this is due to the lack of sessions! This doesn't mean Okta isn't working!).

### Setup to Run the Code

- Ensure you have chamber correctly configured
- Ensure you do not have local environment variables that would replace Okta variables pulled from chamber
  - If you have an env variable that is the same name as the ones pulled, it will overwrite and break

### How to test

Please see the automated auth_test.go file located here: pkg/handlers/authentication/auth_test.go

1. In a private browsing window, open up milmove under the customer, office, or admin domains
2. Click sign-in
3. Accept EULA
4. You should be redirected to Okta for login
5. Create an account if you do not have on yet, otherwise provide your existing credentials
6. After successful login (Also works for account creation), you should be redirected back to milmove
7. You should receive an internal server error due to sessions not being handled yet, however you should now successfully be authenticated into Okta.
8. Your server should have logged an identity error like so:
- `ERROR	authentication/auth.go:870	Error loading Identity.`
9. Verify successful Okta login by navigating to the tenant and checking if you are logged in.
  - Example: https://test-milmove.okta.mil/ 
10. Verify after login if you click "sign-in" again that it will not prompt you for login, it will automatically redirect you back.

### Frontend

- Modified sign-in URL route from login-gov to okta

### Backend

- Code will require truss security review in this file: pkg/cli/auth.go
  - Reason being the word "secret" trips security. This can be put under review when merging to the experimental / staging location. Truss has already stated they would prefer if we pinged them in the PR to their branch.
